contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+  /**
+   * Linearly released balance for a single schedule at the current block timestamp, ignoring any
+   * released amount floor that has been set for the schedule.
+   */
+  /* solhint-disable not-rely-on-time, this is time-dependent */
+  function _linearReleaseAmountAvailable(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 linearAmountAvailable;
+    // slither-disable-next-line timestamp this contract is block time dependent
+
+    if (block.timestamp >= schedule.endTime) {
+      linearAmountAvailable = totalSupply(scheduleId);
+    } else {
+      uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+      // slither-disable-next-line timestamp this contract is block time dependent
+      linearAmountAvailable = block.timestamp < schedule.startTime
+        ? 0
+        : (_scheduleTrueTotal(schedule, scheduleId) *
+          (block.timestamp - schedule.startTime)) / rampTotalTime;
+    }
+    return linearAmountAvailable;
+  }
+
+  /* solhint-enable not-rely-on-time */
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
+    uint256 scheduleId,
+    address account
+  ) private view returns (uint256) {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 scheduleTrueTotal = _scheduleTrueTotal(schedule, scheduleId);
+    uint256 revocableForAccount;
+    // avoid division by 0
+    if (scheduleTrueTotal == 0) {
+      revocableForAccount = 0;
+    } else {
+      uint256 balanceOfAccount = balanceOf(account, scheduleId);
+      uint256 claimedAmountForAccount = schedule.claimedAmountsByAddress[
+        account
+      ];
+      revocableForAccount =
+        ((claimedAmountForAccount + balanceOfAccount) *
+          (totalQuantityToRevoke)) /
+        scheduleTrueTotal;
+    }
+
+    return revocableForAccount;
+  }
+
+  /**
+   * @dev The total amount of released tokens available at the current block timestamp for the schedule.
+   * Takes the maximum of either the calculated linearly released amount based on the schedule parameters,
+   * or the released amount floor, which is set at the current released amount whenever the balance of a
+   * schedule is decreased through revocation or withdrawal.
+   */
+  function _releasedBalanceOfSingleSchedule(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    return
+      MathUpgradeable.max(
+        _linearReleaseAmountAvailable(scheduleId),
+        schedule.releasedAmountFloor
+      );
+  }
+
+  /** Converts an enumerable set into an array */
+  function _enumerableSetToArray(
```

Comment:
is this not just a more gas-inefficent version of `EnumerableSetUpgradeable.values()`

Reply:
oh yeah looks like it is... `.values()` isn't noted anywhere in the OZ documentation but I see it come up on the object, good to know